GVFS/GVFS.UnitTests/Maintenance/LooseObjectStepTests.cs

```diff
@@ -61,21 +63,72 @@ public void LooseObjectsPassTimeRestriction()
+            commands.Count.ShouldEqual(2);
             commands[0].ShouldEqual(PrunePackedCommand);
+            commands[1].ShouldEqual(this.packCommand);
+        }
+
+        [TestCase]
+        public void LooseObjectsLimitPackCount()
+        {
+            this.TestSetup(DateTime.UtcNow.AddDays(-7));
+
+            // Verify with default limit
+            LooseObjectsStep step = new LooseObjectsStep(this.context, requireCacheLock: false, forceRun: false);
+            step.PackLooseObjects(new StreamWriter(new MemoryStream()));
+            step.LooseObjectsPutIntoPackFile.ShouldEqual(3);
+
+            // Verify with limit of 2
+            step.MaxLooseObjectsInPack = 2;
+            step.PackLooseObjects(new StreamWriter(new MemoryStream()));
+            step.LooseObjectsPutIntoPackFile.ShouldEqual(2);
+        }
+
+        [TestCase]
+        public void SkipInvalidLooseObjects()
+        {
+            this.TestSetup(DateTime.UtcNow.AddDays(-7));
+
+            // Verify with valid Objects 
+            LooseObjectsStep step = new LooseObjectsStep(this.context, requireCacheLock: false, forceRun: false);
+            step.PackLooseObjects(new StreamWriter(new MemoryStream()));
+            step.LooseObjectsPutIntoPackFile.ShouldEqual(3);
+            this.tracer.RelatedErrorEvents.Count.ShouldEqual(0);
+            this.tracer.RelatedWarningEvents.Count.ShouldEqual(0);
+
+            // Write an ObjectId file with an invalid name
+            this.context.FileSystem.WriteAllText(Path.Combine(this.context.Enlistment.GitObjectsRoot, "AA", "NOT_A_SHA"), string.Empty);
+
+            // Verify it wasn't added and a warning exists
+            step.PackLooseObjects(new StreamWriter(new MemoryStream()));
+            step.LooseObjectsPutIntoPackFile.ShouldEqual(3);
+            this.tracer.RelatedErrorEvents.Count.ShouldEqual(0);
+            this.tracer.RelatedWarningEvents.Count.ShouldEqual(1);
         }
 
         [TestCase]
         public void LooseObjectsCount()
         {
-            this.TestSetup(DateTime.Now.AddDays(-7));
+            this.TestSetup(DateTime.UtcNow.AddDays(-7));
 
             LooseObjectsStep step = new LooseObjectsStep(this.context, requireCacheLock: false, forceRun: false);
             int count = step.CountLooseObjects();
 
             count.ShouldEqual(3);
         }
 
+        [TestCase]
+        public void LooseObjectId()
+        {
+            this.TestSetup(DateTime.UtcNow.AddDays(-7));
+
+            LooseObjectsStep step = new LooseObjectsStep(this.context, requireCacheLock: false, forceRun: false);
+            string directoryName = "AB";
+            string fileName = "830bb79cd4fadb2e73e780e452dc71db909001";
+            string objectId = step.GetLooseObjectId(directoryName, Path.Combine(this.context.Enlistment.GitObjectsRoot, directoryName, fileName));
+            objectId.ShouldEqual(directoryName + fileName);
+        }
+
         private void TestSetup(DateTime lastRun)

```

Comment:
I would have this return the string for the pack command instead of storing it in at member variable so one test can't affect another and it is easier in the test to know where the `packCommand` is getting set.

Reply:
We set quite a few variables values in TestSetup.  Originally i had them as out parameters but that was agreed to be ugly.  I think the user can determine where the packCommand is set.